### PR TITLE
Added ShuffleWithinSegments

### PR DESCRIPTION
**Pull request by Erlemar**
Shuffling tokens within segments.

    The token sequence is split into segments of the same label; for each segment, a decision is made randomly: to shuffle it or not.
    In case of a positive decision, the tokens within each segment are shuffled.

The source: https://arxiv.org/abs/2010.11683

### CODE REVIEW

**Comment by Erlemar** [comment]
SuspectingParaphraser doesn't pass the tests. What should I do about it.

File: transformations/shuffle_within_segments/test.json [hunk 1]
```diff
@@ -0,0 +1,71 @@
+{
```
Comment (sebastianGehrmann) [inline transformations/shuffle_within_segments/test.json:1]: The method signature expects Lists but your test cases use strings.

**Review by sebastianGehrmann — CHANGES_REQUESTED** [review]
I suspect your tests fails since your test cases do not conform to the expect data format (list vs. string). I can take another look when this is fixed.

File: transformations/shuffle_within_segments/test.json (hunk 1 above)
Comment (Erlemar) [reply transformations/shuffle_within_segments/test.json:1]: The problem is that tests for my transformation (ShuffleWithinSegments) pass successfully, tests fail for a different transformation (SuspectingParaphraser).
As for the signature, I have followed the example of some other TEXT_TAGGING transformers like https://github.com/GEM-benchmark/NL-Augmenter/tree/main/transformations/longer_location_ner

**Comment by marco-digio** [comment]
Hi @Erlemar 
I was assigned as one of your reviewers. 
Thanks for this contribution! 
I this that this transformation is very useful and clear. 

Here's my general review:
Correctness: All checks have failed, please fix this by looking at other PRs
Interface: I think you have chosen the correct interface(s).
Applicable Tasks & Keywords: I think you have chosen the relevant task. Please add keywords as explained [here](https://github.com/GEM-benchmark/NL-Augmenter/blob/main/docs/keywords.md)
Specificity: The change is general.
Novelty: This transformation is not yet implemented in NL-Augmenter.
Adding New Libraries: No new libraries appear to be added, and there is no requirements.txt file.
Description: The README is clear and explains what the transformation is aiming to do.
Data and code source: References clearly indicated in README
Test Cases: 6 test cases have been added.
Evaluating Robustness: No robustness evaluation has been performed.
Languages other than English: Yes, but it is required as input a TAG sequence

**Comment by Erlemar** [comment]
Hello @marco-digio 

Thank you for your review!

Correctness:

The problem is that tests for my transformation (ShuffleWithinSegments) pass successfully, tests fail for a different transformation (SuspectingParaphraser). I'm not sure what should I do about it.

Applicable Tasks & Keywords: I have added the keywords, but I'm not sure that I have selected all of them correctly. Here is my reasoning:
"lexical" - the transformation substitutes different words.
"word-order" -  the order of the words is changed.
"rule-based" - the transformation uses heuristics.
"unnaturally-written", "unnatural-sounding" - I'm not sure about these keywords: this transformation randomly shuffles words within separate entities, as a result, the order of the words could be unnatural. This perturbed order could sound unnatural too. This is why I chose both these keywords.
"possible-meaning-preserving" - due to the changed word order, the meaning of the sentence could change.
"high-precision" - the transformation generates what it intends to generate.
"low-coverage" - the transformation works only for TAG sequences.
"low-generations" - The transformation generates few outputs for a single input.

Evaluating Robustness: I have added the results of the evaluation check.

**Comment by Erlemar** [comment]
I have pushed the changes from https://github.com/GEM-benchmark/NL-Augmenter/pull/288 now the tests should pass.

**Review by samsontmr — COMMENTED** [review]
Hi @Erlemar! It seems tests are still failing

**Comment by Erlemar** [comment]
Hi @samsontmr @marco-digio the tests have passsed successfully.

File: transformations/shuffle_within_segments/test.json (hunk 1 above)
Comment (Erlemar) [reply transformations/shuffle_within_segments/test.json:1]: The tests work now.

**Review by Erlemar — COMMENTED** [review]
The tests work now.

File: transformations/shuffle_within_segments/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+Dataset({
+    features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
+    num_rows: 691
+})
+
+The average accuracy on a subset of conll2003 = 81.3640468775567
+
+The average accuracy on its perturbed set = 70.13742999132477
+
+## Previous Work
```
Comment (samsontmr) [inline transformations/shuffle_within_segments/README.md:31]: Could you replace this bibtex with the one from [the ACL anthology](https://aclanthology.org/2020.coling-main.343)?

File: transformations/shuffle_within_segments/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+## What type of a transformation is this?
+This transformation shuffles the tokens in the sentence without breaking the sequences of the labels.
+
+## What tasks does it intend to benefit?
+This perturbation would benefit all tasks which have a sentence/paragraph/document as input like text classification, 
+text generation, and most importantly, a tagging task.
+This would help augment data for a NER task by keeping the labels still aligned.
+
+## What are the limitations of this transformation?
+It is possible that shuffling could change the meaning of the sentence or have negative effect for NLU tasks.
```
Comment (samsontmr) [inline transformations/shuffle_within_segments/README.md:16]: Another effect is that changing the word order could change the entity's "meaning"

**Review by samsontmr — COMMENTED** [review]
Looks to mostly be in good shape!

File: transformations/shuffle_within_segments/README.md (hunk 2 above)
Comment (Erlemar) [reply transformations/shuffle_within_segments/README.md:31]: I have replaced this bibtex.

File: transformations/shuffle_within_segments/README.md (hunk 3 above)
Comment (Erlemar) [reply transformations/shuffle_within_segments/README.md]: Updated the description.

File: transformations/shuffle_within_segments/README.md [hunk 4]
```diff
@@ -13,7 +13,7 @@ text generation, and most importantly, a tagging task.
 This would help augment data for a NER task by keeping the labels still aligned.
 
 ## What are the limitations of this transformation?
-It is possible that shuffling could change the meaning of the sentence or have negative effect for NLU tasks.
+It is possible that shuffling could change the meaning of the sentence, the meaning the entity itself or have negative effect for NLU tasks.
```
Comment (samsontmr) [inline transformations/shuffle_within_segments/README.md:16]: the meaning **_of_** the entity itself or have **_a_** negative effect **_on_** NLU tasks

Comment (Erlemar) [reply transformations/shuffle_within_segments/README.md]: I have fixed the typos.

File: transformations/shuffle_within_segments/transformation.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                "low-coverage",
+                "low-generations"]
+
+    def __init__(self, seed=0, max_outputs=1):
+        super().__init__(seed, max_outputs=max_outputs)
+
+    def generate(
+            self, token_sequence: List[str], tag_sequence: List[str]
+    ) -> List[Tuple[List[str], List[str]]]:
+        random.seed(self.seed)
```
Comment (samsontmr) [inline transformations/shuffle_within_segments/transformation.py:38]: I believe this is no longer needed

Comment (Erlemar) [reply transformations/shuffle_within_segments/transformation.py]: I have removed it now.

**Review by sebastianGehrmann — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
Thanks @Erlemar for the hard work and everyone for their comments.

**Pull request merged**